src/Terminal/ui/TerminalRoot.tsx

```diff
@@ -112,20 +112,136 @@ export function TerminalRoot({ terminal, router, player }: IProps): React.ReactE
+        const r = ir <= 0 ? 0 : 55 + ir * 40
+        const g = ig <= 0 ? 0 : 55 + ig * 40
+        const b = ib <= 0 ? 0 : 55 + ib * 40
+        return `rgb(${r}, ${g}, ${b})`
+      }
+      if((232 <= code) && (code < 256)){
+        // x32 greyscale
+        const base = code - 232
+        const grey = base * 10 + 8
+        return `rgb(${grey}, ${grey}, ${grey})`
+      }
+      // shouldn't get here (under normal circumstances), but just in case
+      return 'initial'
     }
-    if (s === "warn") {
-      return classes.warning;
+    type styleKey = "fontWeight" | "textDecoration" | "color" | "backgroundColor"
+    const style : {
+      fontWeight?: string;
+      textDecoration?: string;
+      color?: string;
+      backgroundColor?: string;
+    } = {}
+
+    if(code !== null){ /* eslint-disable yoda */
+      const codeParts = code.split(';').map((p) => parseInt(p)).filter((p,i,arr) =>
+        // If the sequence is 38;5 (x256 foreground color) or 48;5 (x256 background color),
+        //  filter out the 5 so the next codePart after {38,48} is the color code.
+        (p!=5) || i==0 || ((arr[i-1] != 38)&&(arr[i-1] != 48))
+      )
+
+      let nextStyleKey:styleKey|null = null;
+      codeParts.forEach(codePart => {
+        if(nextStyleKey !== null){
+          style[nextStyleKey] = ansi2rgb(codePart)
+          nextStyleKey = null
+        }
+        // Decorations
+        else if(codePart == 1){
+          style.fontWeight = 'bold'
+        }
+        else if(codePart == 4){
+          style.textDecoration = 'underline'
+        }
+        // Forground Color (x8)
+        else if((30 <= codePart)&&(codePart < 38)){
+          if(COLOR_MAP[codePart%10]){
+            style.color = COLOR_MAP[codePart%10]
+          }
+        }
+        // Background Color (x8)
+        else if((40 <= codePart)&&(codePart < 48)){
+          if(COLOR_MAP[codePart%10]){
+            style.backgroundColor = COLOR_MAP[codePart%10]
+          }
+        }
+        // Forground Color (x256)
+        else if(codePart == 38){
+          nextStyleKey = 'color'
+        }
+        // Background Color (x256)
+        else if(codePart == 48){
+          nextStyleKey = 'backgroundColor'
+        }
+      })
     }
+    return style
+  }
 
-    return classes.primary;
+  function parseOutputText(item: Output): React.ReactElement {
+    // Some things, oddly, can cause item.text to not be a string (e.g. expr from the CLI), which
+    //  causes .matchAll to throw.  Ensure it's a string immediately
+    const parts = []
+    if(typeof item.text === 'string'){
+      // eslint-disable-next-line no-control-regex
+      const ANSI_ESCAPE = new RegExp('\u{001b}\\[(?<code>.*?)m', 'ug');
+      const matches = [null, ...item.text.matchAll(ANSI_ESCAPE), null]
+      if(matches.length > 2){
+        matches.slice(0, -1).forEach((m,i) => {
+          const n = matches[i+1]
+          if(m&&m.index!==undefined&&m.groups!==undefined){
```

Comment:
It would be better to fail early than to wrap the entire contents within an if statement, to reduce the amount of indentation. Eg;

```js
if (!m || m.index === undefined || m.groups === undefined) {
  return;
}

const startIndex = m ? m.index + m[0].length : 0
const stopIndex  = n ? n.index               : item.text.length
const partText   = item.text.slice(startIndex, stopIndex)
parts.push({code:m.groups.code, text:partText})
```